File: 0x43-线段树/acwing245.py
def readToInt():
    return [int(n) for n in input().split()]

class TreeNode(object):
    def __init__(self, left, right, sum=0, lmax=0, rmax=0, max=0):
        self.left = left
        self.right = right
        self.sum = sum
        self.lmax = lmax
        self.rmax = rmax
        self.max = max

class SegementTree(object):
    def __init__(self, objects):
        self.nodes = [TreeNode(0, 0) for _ in range(4 * len(objects))]
        self.objects = objects

    def build(self, i, left, right):
        self.nodes[i].left = left
        self.nodes[i].right = right
        if left == right:
            self.nodes[i].sum = self.objects[left - 1]
            self.nodes[i].lmax, self.nodes[i].rmax, self.nodes[i].max = self.nodes[i].sum, self.nodes[i].sum, self.nodes[i].sum
            return
        mid = (left + right) // 2
        self.build(2 * i, left, mid)
        self.build(2 * i + 1, mid + 1, right)
        self.nodes[i].sum = self.nodes[2 * i].sum + self.nodes[2 * i + 1].sum
        self.nodes[i].lmax = max(self.nodes[2 * i].lmax, self.nodes[2 * i].sum + self.nodes[2 * i + 1].lmax)
        self.nodes[i].rmax = max(self.nodes[2 * i + 1].rmax, self.nodes[2 * i + 1].sum + self.nodes[2 * i].rmax)
        self.nodes[i].max = max(self.nodes[2 * i].max, self.nodes[2 * i + 1].max, self.nodes[2 * i].rmax + self.nodes[2 * i + 1].lmax)

    def change(self, i, x, y):
        if self.nodes[i].left == self.nodes[i].right:
            self.nodes[i].sum, self.nodes[i].lmax, self.nodes[i].rmax, self.nodes[i].max = y, y, y, y
            return
        mid = (self.nodes[i].left + self.nodes[i].right) // 2
        if x <= mid:
            self.change(2 * i, x, y)
        else:
            self.change(2 * i + 1, x, y)
        self.nodes[i].sum = self.nodes[2 * i].sum + self.nodes[2 * i + 1].sum
        self.nodes[i].lmax = max(self.nodes[2 * i].lmax, self.nodes[2 * i].sum + self.nodes[2 * i + 1].lmax)
        self.nodes[i].rmax = max(self.nodes[2 * i + 1].rmax, self.nodes[2 * i + 1].sum + self.nodes[2 * i].rmax)
        self.nodes[i].max = max(self.nodes[2 * i].max, self.nodes[2 * i + 1].max, self.nodes[2 * i].rmax + self.nodes[2 * i + 1].lmax)

    def ask(self, i, x, y):
        if x <= self.nodes[i].left and self.nodes[i].right <= y:
            return self.nodes[i]
        mid = (self.nodes[i].left + self.nodes[i].right) // 2
        ans = TreeNode(x, y)
        if mid < x:
            ans = self.ask(2 * i + 1, x, y)
        elif y <= mid:
            ans = self.ask(2 * i, x, y)
        else:
            left, right = self.ask(2 * i, x, y), self.ask(2 * i + 1, x, y)
            ans.sum = left.sum + right.sum
            ans.lmax = max(left.lmax, left.sum + right.lmax)
            ans.rmax = max(right.rmax, right.sum + left.rmax)
            ans.max = max(left.max, right.max, left.rmax + right.lmax)
        return ans

NM = readToInt()
nums = readToInt()
stree = SegementTree(nums)
stree.build(1, 1, len(nums))
for _ in range(NM[1]):
    query = readToInt()
    if query[0] == 1:
        print(stree.ask(1, min(query[1], query[2]), max(query[1], query[2])).max)
    else:
        stree.change(1, query[1], query[2])

File: 0x43-线段树/Acwing245.java
import java.util.Scanner;

public class Acwing245 {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int N = scan.nextInt();
        int M = scan.nextInt();
        SegmentTree stree = new SegmentTree(N, scan);
        stree.build(1, 1, N);
        scan.nextLine();
        for (int i = 0; i < M; i++) {
            String[] nums = scan.nextLine().split(" ");
            int x = Integer.valueOf(nums[1]), y = Integer.valueOf(nums[2]);
            if(nums[0].equals("1")){
                int ans = stree.query(1, Math.min(x, y), Math.max(x, y));
                System.out.println(ans);
            }else{
                stree.change(1, x, y);
            }
        }
    }
}

class SegmentTree{
    private class Node{
        int l, r, sum, lmax, rmax, max;
        public Node(int l, int r, int sum, int lmax, int rmax, int max){
            this.l = l; this.r = r; this.sum = sum; this.lmax = lmax; this.rmax = rmax; this.max = max;
        }
        public Node(int l, int r){
            this(l, r, 0, 0, 0, 0);
        }
    }
    Node[] nodes;
    Scanner scan;
    public SegmentTree(int N, Scanner scan){
        this.nodes = new Node[4 * N];
        this.scan = scan;
    }

    public void build(int i, int l, int r){
        nodes[i] = new Node(l, r);
        if(l == r){
            int v = scan.nextInt();
            nodes[i].sum = nodes[i].lmax = nodes[i].rmax = nodes[i].max = v;
            return;   
        }
        int mid = (l + r) >> 1;
        build(i << 1, l, mid);
        build(i << 1 | 1, mid + 1, r);
        push(nodes[i], nodes[i << 1], nodes[i << 1 | 1]);
    }

    public void change(int i, int x, int y){
        if(nodes[i].l == nodes[i].r){
            nodes[i].sum = nodes[i].lmax = nodes[i].rmax = nodes[i].max = y;
            return;
        }
        int mid = (nodes[i].l + nodes[i].r) >> 1;
        if(mid < x){
            change(i << 1 | 1, x, y);
        }else{
            change(i << 1, x, y);
        }
        push(nodes[i], nodes[i << 1], nodes[i << 1 | 1]);
    }

    public int query(int i, int l, int r){
        return ask(i, l, r).max;
    }

    private Node ask(int i, int l, int r){
        if(l <= nodes[i].l && nodes[i].r <= r){
            return nodes[i];
        }
        int mid = (nodes[i].l + nodes[i].r) >> 1;
        if(mid < l){
            return ask(i << 1 | 1, l, r);
        }
        if(r <= mid){
            return ask(i << 1, l, r);
        }
        Node root = new Node(l, r);
        Node left = ask(i << 1, l, r);
        Node right = ask(i << 1 | 1, l, r);
        push(root, left, right);
        return root;
    }

    private void push(Node root, Node left, Node right){
        root.sum = left.sum + right.sum;
        root.lmax = Math.max(left.lmax, left.sum + right.lmax);
        root.rmax = Math.max(right.rmax, right.sum + left.rmax);
        root.max = Math.max(Math.max(left.max, right.max), left.rmax + right.lmax);
    }
}